### PR TITLE
Standalone POS

### PR DESCRIPTION
**Pull request by Talej**
# Description

Implements standalone POS functionality. 

- Changes the enable square POS setting to allow enable Square or Standalone mode.
- Under POS settings adds support for managing categories and products.
- Products can be priced in sats or fiat.
- Adds a standalone POS order creation screen and allows invoicing through the existing payment flow.

** Note: ** I haven't yet tested this on an Android device. Currently without a working device

This pull request is categorized as a:

- [X] New feature
- [ ] Bug fix
- [ ] Code refactor
- [ ] Configuration change
- [ ] Locales update
- [ ] Quality assurance 
- [ ] Other

## Checklist
- [ ] I’ve run `yarn run tsc` and made sure my code compiles correctly
- [X] I’ve run `yarn run lint` and made sure my code didn’t contain any problematic patterns
- [X] I’ve run `yarn run prettier` and made sure my code is formatted correctly
- [X] I’ve run `yarn run test` and made sure all of the tests pass

## Testing

If you modified or added a utility file, did you add new unit tests?

- [ ] No, I’m a fool
- [ ] Yes
- [X] N/A

I have tested this PR on the following platforms (please specify OS version and phone model/VM):

- [ ] Android
- [X] iOS

I have tested this PR with the following types of nodes (please specify node version and API version where appropriate):

- [X] Embedded LND
- [ ] LND (REST)
- [ ] LND (Lightning Node Connect)
- [ ] Core Lightning (c-lightning-REST)
- [ ] LndHub
- [ ] [DEPRECATED] Core Lightning (Spark)
- [ ] [DEPRECATED] Eclair

### Locales
- [X] I’ve added new locale text that requires translations
- [ ] I’m aware that new translations should be made on the Zeus [Transfix page](https://app.transifex.com/ZeusLN/zeus/) and not directly to this repo

### Third Party Dependencies and Packages

- [ ] Contributors will need to run `yarn` after this PR is merged in
- [ ] 3rd party dependencies have been modified:
    * verify that `package.json` and `yarn.lock` have been properly updated
    * verify that dependencies are installed for both iOS and Android platforms

### Other:

- [ ] Changes were made that require an update to the README
- [ ] Changes were made that require an update to onboarding

### CODE REVIEW

**Review by shubhamkmr04 — COMMENTED** [review]
Simulator Screenshot - iPhone 14 Plus - 2023-10-12 at 18 45 54
Getting this warning as soon as we open the POS settings

**Review by shubhamkmr04 — COMMENTED** [review]
https://github.com/ZeusLN/zeus/assets/50761978/8ed84874-13d6-4286-ac6f-a62bb6664da3

Not able to change the unit to 'sats'

**Review by shubhamkmr04 — COMMENTED** [review]
If the Category gets deleted, the product belonging to that Category should also be deleted.

**Review by shubhamkmr04 — COMMENTED** [review]
https://github.com/ZeusLN/zeus/assets/50761978/6864f619-9585-46e6-894f-b8d23550f00b

So we add our product to checkout and pay, then delete it from the settings, come to the homepage, and still can see our checked-out price belonging to our deleted product.

**Comment by kaloudis** [comment]
> Screen.Recording.2023-10-12.at.10.08.11.PM.mov
> 
> We should have a counter here on this view to know how many products we have selected so far and maybe a way to remove it, too. @kaloudis, what do you think about this?

+1 on having a counter

**Comment by kaloudis** [comment]
> Screen.Recording.2023-10-12.at.10.13.21.PM.mov
> 
> So we add our product to checkout and pay, then delete it from the settings, come to the homepage, and still can see our checked-out price belonging to our deleted product.

Deleting a product should probably trigger a cart clear

**Comment by Talej** [comment]
> Screen.Recording.2023-10-12.at.10.08.11.PM.mov 
> We should have a counter here on this view to know how many products we have selected so far and maybe a way to remove it, too. @kaloudis, what do you think about this?

I have added an item counter in the "Charge" button.

Other thing I could think to do is add - (n) + in each of the product components. Might not be super user friendly though IMO.

**Comment by Talej** [comment]
> Simulator Screenshot - iPhone 14 Plus - 2023-10-12 at 18 45 54 Getting this warning as soon as we open the POS settings

This one would've already been in there previously but I have sorted it out by dropping the FlatList

**Comment by Talej** [comment]
All of these items and a couple more things I found should be sorted out now.

**Comment by shubhamkmr04** [comment]
> > Screen.Recording.2023-10-12.at.10.08.11.PM.mov
> > We should have a counter here on this view to know how many products we have selected so far and maybe a way to remove it, too. @kaloudis, what do you think about this?
> 
> I have added an item counter in the "Charge" button.
> 
> Other thing I could think to do is add - (n) + in each of the product components. Might not be super user friendly though IMO.

Yup, I understood. @kaloudis, should we do that, or the overall counter would be fine

**Review by shubhamkmr04 — COMMENTED** [review]
Should we also remove open orders of deleted products?🤔

**Review by shubhamkmr04 — COMMENTED** [review]
https://github.com/ZeusLN/zeus/assets/50761978/bba7ec0d-0144-4d77-9da0-5a80b53577f1


We cannot save a product to the first category in the list; it goes into the uncategorized products. It only happens in the android.

**Review by shubhamkmr04 — COMMENTED** [review]
On the **ProductDetails** view, In the delete button, we should have our props                                 `containerStyle` and `titleStyle` set to `themeColor('delete')` to maintain consistency with other views.

**Comment by shubhamkmr04** [comment]
> > Screen.Recording.2023-10-12.at.10.13.21.PM.mov
> > So we add our product to checkout and pay, then delete it from the settings, come to the homepage, and still can see our checked-out price belonging to our deleted product.
> 
> Deleting a product should probably trigger a cart clear

Would it be better to subtract that deleted product's price instead of clearing the whole cart?

**Comment by Talej** [comment]
> > > Screen.Recording.2023-10-12.at.10.13.21.PM.mov
> > > So we add our product to checkout and pay, then delete it from the settings, come to the homepage, and still can see our checked-out price belonging to our deleted product.
> > 
> > 
> > Deleting a product should probably trigger a cart clear
> 
> Would it be better to subtract that deleted product's price instead of clearing the whole cart?

I think it would be a pretty rare condition that the user is editing/deleting products at the same time as actually making sales so I feel like just clearing the whole cart should be sufficient. What do you think?

**Comment by Talej** [comment]
> Should we also remove open orders of deleted products?🤔

I feel like maybe this would lead to confusion and it should probably be left to the user to either hide/cancel the order or create a new one.

**Comment by Talej** [comment]
- Fixed category delete issue with products
- Fixed issue with selecting category for product on android
- Updated styling for delete buttons

Ready for another round!

**Comment by kaloudis** [comment]
> > > > Screen.Recording.2023-10-12.at.10.13.21.PM.mov
> > > > So we add our product to checkout and pay, then delete it from the settings, come to the homepage, and still can see our checked-out price belonging to our deleted product.
> > > 
> > > 
> > > Deleting a product should probably trigger a cart clear
> > 
> > 
> > Would it be better to subtract that deleted product's price instead of clearing the whole cart?
> 
> I think it would be a pretty rare condition that the user is editing/deleting products at the same time as actually making sales so I feel like just clearing the whole cart should be sufficient. What do you think?

This should suffice

**Comment by kaloudis** [comment]
> > Should we also remove open orders of deleted products?🤔
> 
> I feel like maybe this would lead to confusion and it should probably be left to the user to either hide/cancel the order or create a new one.

can go either way here. I don't think this is a deal breaker either way. maybe something that gets iterated on

File: views/POS/ProductDetails.tsx [hunk 1]
```diff
@@ -0,0 +1,504 @@
...
+                }));
+                let categoryOptions: any[] = [
+                    {
+                        key: 'Uncategorized',
+                        value: '',
+                        translateKey: 'pos.views.Wallet.PosPane.uncategorized'
+                    }
+                ];
+                categoryOptions = categoryOptions.concat(...mappedCategories);
+                console.log('options', categoryOptions);
```
Comment (shubhamkmr04) [inline views/POS/ProductDetails.tsx:97]: Do we need this console, or was it just for debugging?

Comment (Talej) [reply views/POS/ProductDetails.tsx]: Was just for debugging. Have removed now

**Comment by Talej** [comment]
> Screen.Recording.2023-10-23.at.4.25.41.PM.mov 
> The screen returns to the top when I select any product, which might affect the user's experience as they would have to scroll down again and again if we have so many products listed.

Having the Charge button up in the Tab Navigator was triggering re-render on every touch. I have restructured it so that doesn't happen any more.

Unfortunately means that I can't use the Tab Navigator space now so we've lost a bit of screen real estate

**Comment by shubhamkmr04** [comment]
Simulator Screenshot - iPhone 14 Plus - 2023-10-27 at 19 09 20
@kaloudis, does the button look good in here?

**Review by shubhamkmr04 — COMMENTED** [review]
https://github.com/ZeusLN/zeus/assets/50761978/54fddd31-ffa6-4045-a0e7-40b83b608825

The amount is stuck on the price of first product I selected.

**Comment by Talej** [comment]
> Screen.Recording.2023-10-27.at.7.24.49.PM.mov 
> The amount is stuck on the price of first product I selected.

Not sure how I missed that one. Nice catch. Fixed now!

**Review by shubhamkmr04 — COMMENTED** [review]
I think we shouldn't clear the cart when returning from another screen.

**Review by shubhamkmr04 — COMMENTED** [review]
Simulator Screenshot - iPhone 14 Plus - 2023-10-30 at 14 43 57
Simulator Screenshot - iPhone 14 Plus - 2023-10-30 at 14 43 45
I am getting a weird UI behavior in IOS where the category heading gets stuck on the top until it is scrolled completely.

**Review by shubhamkmr04 — COMMENTED** [review]
The latest orders should be listed on the top, just like we get the latest transactions first in the activity view.

**Review by shubhamkmr04 — COMMENTED** [review]
https://github.com/ZeusLN/zeus/assets/50761978/3a831eef-2793-4268-84ab-eee0fc76f011

Not able to edit category.

**Review by shubhamkmr04 — COMMENTED** [review]
What is the context of the checkbox `Active` while creating the product?

**Comment by Talej** [comment]
> Maybe we shouldn't clear the cart when returning from another screen, only when the product is deleted.

It definitely shouldn't be clearing the cart just by changing screens. I have tried to reproduce this a few different ways but it doesn't seem to happen here.

Can you provide any more details?

**Comment by Talej** [comment]
> Simulator Screenshot - iPhone 14 Plus - 2023-10-30 at 14 43 57 Simulator Screenshot - iPhone 14 Plus - 2023-10-30 at 14 43 45 I am getting a weird UI behavior in IOS where the category heading gets stuck on the top until it is scrolled completely.

This is just the default SectionList behaviour for iOS. I should be able to disable it

**Comment by Talej** [comment]
> The latest orders should be listed on the top, just like we get the latest transactions first in the activity view.

I will update to sort these

**Comment by Talej** [comment]
> What is the context of the checkbox `Active` while creating the product?

The main product POS screen should only be showing `Active` products. Will update it

**Comment by shubhamkmr04** [comment]
> > Maybe we shouldn't clear the cart when returning from another screen, only when the product is deleted.
> 
> It definitely shouldn't be clearing the cart just by changing screens. I have tried to reproduce this a few different ways but it doesn't seem to happen here.
> 
> Can you provide any more details?


https://github.com/ZeusLN/zeus/assets/50761978/0232c8d3-8c75-4777-a277-bd9858627ee8

**Review by shubhamkmr04 — COMMENTED** [review]
Also, please check the UI for other themes as well. Maybe we need to make `background: none` in some places to make it consistent.

**Comment by Talej** [comment]
> Also, please check the UI for other themes as well. Maybe we need to make `background: none` in some places to make it consistent.

For this one I have removed the backgroundColor on the section headers. Haven't made any other changes but I'm open to suggestions.

**Comment by Talej** [comment]
Latest changes have been pushed that should cover everything to date. Let me know if there is anything else I've missed.

**Review by shubhamkmr04 — COMMENTED** [review]
Let's clear the cart in case we make a selected product inactive.

**Review by shubhamkmr04 — COMMENTED** [review]
Also, could you remove the search in all the views when we have nothing to show?

File: views/Settings/PointOfSale.tsx [hunk 2]
```diff
@@ -415,41 +414,35 @@ export default class PointOfSale extends React.Component<
...
-                                }
-                            />
-                        )}
+                        {posEnabled !== PosEnabled.Disabled &&
+                            LIST_ITEMS.map((item, index) => (
+                                <ListItem
+                                    containerStyle={{
+                                        borderBottomWidth: 0,
+                                        backgroundColor:
+                                            themeColor('background')
```
Comment (shubhamkmr04) [inline views/Settings/PointOfSale.tsx:423]: Use `backgroundColor: 'none` here.

**Comment by Talej** [comment]
In theory everything should be rebased properly and latest feedback all implemented!

**Comment by kaloudis** [comment]
Looking great!

**Comment by kaloudis** [comment]
Alignment is a little off here on my Android simulator. Could we potentially use the `AmountInput` component here?

Screenshot 2023-12-11 at 15 13 13

**Comment by kaloudis** [comment]
Might be better fit for another PR but I think we should consider adding a keypad pane to the POS for merchants that want to ring up a quick payment without specifying items. Perhaps should be a POS setting.

We should also remove the easter egg to reveal the POS in Settings. It should now be available for everyone to use.

**Comment by Talej** [comment]
> Might be better fit for another PR but I think we should consider adding a keypad pane to the POS for merchants that want to ring up a quick payment without specifying items. Perhaps should be a POS setting.
> 
> We should also remove the easter egg to reveal the POS in Settings. It should now be available for everyone to use.

Have removed the easter egg check on the menu directly but left all the easter egg code in place just so it can be used for other options going forward.

Let's do the keypad as a separate PR

**Comment by kaloudis** [comment]
Screenshot 2023-12-18 at 19 42 36

hitting an issue here where the `AmountInput` component is pre-filled with value `undefined` and I cannot change the amount

**Comment by Talej** [comment]
> Screenshot 2023-12-18 at 19 42 36
> hitting an issue here where the `AmountInput` component is pre-filled with value `undefined` and I cannot change the amount

creating a new product or editing an existing?

**Comment by kaloudis** [comment]
> > Screenshot 2023-12-18 at 19 42 36
> > hitting an issue here where the `AmountInput` component is pre-filled with value `undefined` and I cannot change the amount
> 
> creating a new product or editing an existing?

create a new product

**Comment by kaloudis** [comment]
Looking good now! I've also double checked the structure to ensure that fiat prices are recorded properly.

**Review by shubhamkmr04 — COMMENTED** [review]
The amount is prefilled with 0 that we can't remove 

Screenshot 2023-12-20 at 12 11 14 PM


Screenshot 2023-12-20 at 12 11 48 PM

**Comment by Talej** [comment]
> The amount is prefilled with 0 that we can't remove
> 
> Screenshot 2023-12-20 at 12 11 14 PM Screenshot 2023-12-20 at 12 11 48 PM

Fixed

**Comment by shubhamkmr04** [comment]
Looks fine now.

**Comment by kaloudis** [comment]
Code looks good. A few things missing though:

- Ability (or option) to go to order overview before generating invoice (allows user to set tip on this screen and see the total w/ tax) [you can go to this view if you don't pay and click on an order under the Open tab)
- Ability to set tax %
- Ability to disable tips

LMK what you think @Talej. Open to shipping it in beta1 as is and breaking up follow up work in other PRs.

**Comment by kaloudis** [comment]
nit: let's also put the Standalone option above Square in the selection menu

**Comment by kaloudis** [comment]
I've made the suggested changes myself to the PR:

- Standalone placed above Square in the mode selection dropdown
- Charging an order now takes you to an itemized breakdown instead of directly to the invoice (let's give some thought into if we want to make this configurable)
- Disable tip and conf preferences are now displayed when on the POS settings when in Standalone mode
- New tax percentage setting added to POS settings when in Standalone mode

I also noticed a bug with how itemized products were displayed on the `Order view`: amount displayed was 100x lower. Change can be found here: https://github.com/ZeusLN/zeus/pull/1722/commits/ad4927256c5c34d861fd93567c80be67c2e4d99d#diff-7ffe38a0457034ead72df14de5a05813a6c8a8945e64551040568d72e81a7d5eR330 Just need a sanity check here, but it appears to be rendering properly now.

**Comment by Talej** [comment]
> I've made the suggested changes myself to the PR:
> 
> * Standalone placed above Square in the mode selection dropdown
> * Charging an order now takes you to an itemized breakdown instead of directly to the invoice (let's give some thought into if we want to make this configurable)
> * Disable tip and conf preferences are now displayed when on the POS settings when in Standalone mode
> * New tax percentage setting added to POS settings when in Standalone mode
> 
> I also noticed a bug with how itemized products were displayed on the `Order view`: amount displayed was 100x lower. Change can be found here: [ad49272#diff-7ffe38a0457034ead72df14de5a05813a6c8a8945e64551040568d72e81a7d5eR330](https://github.com/ZeusLN/zeus/commit/ad4927256c5c34d861fd93567c80be67c2e4d99d#diff-7ffe38a0457034ead72df14de5a05813a6c8a8945e64551040568d72e81a7d5eR330) Just need a sanity check here, but it appears to be rendering properly now.

This all checks out and looks okay to me!

**Review by shubhamkmr04 — COMMENTED** [review]
tACK

**Pull request merged**